text/0000-write-pointers.md

```diff
@@ -0,0 +1,463 @@
+struct FrameBuffer( ... );
+
+impl FrameBuffer {
+    /// initializes the FrameBuffer in place
+    fn new(&uninit self) { ... }
+
+    /// gets a write only refernce to pixel at position (row, col)
+    fn write_to_pixel(&mut self, row: usize, col: usize) -> &out Rgb {
+         ...
+    }
+}
+```
+This could be used like this
+```Rust
+let buffer;
+FrameBuffer::new(&uninit buffer);
+
+*buffer.write_to_pixel(0, 0) = Rgb(50, 50, 255);
+*buffer.write_to_pixel(10, 20) = Rgb(0, 250, 25);
+/// ...
+```
+
+**Note:** `Rgb` is `Copy`, if it wasn't we could not gaurentee that we can safely overwrite it
+
+## Constructors and Direct Initialization
+
+Using `&uninit` we can create constructors for Rust!
+```Rust
+struct Rgb(u8, u8 ,u8);
+
+impl Rgb {
+    fn init(&uninit self, r: u8, g: u8, b: u8) {
+        self.0 = r;
+        self.1 = g;
+        self.2 = b;
+    }
+}
+
+let color: Rgb;
+color.init(20, 23, 255);
+```
+
+and we can do direct initialization
+```Rust
+impl<T> Vec<T> {
+    pub fn emplace_back(&mut self) -> &uninit T {
+        ... // magic to allocate space and create reference
+    }
+}
+```
+
+and maintain write-only buffers
+```Rust
+struct WriteOnly([u8; 1024]);
+
+impl WriteOnly {
+    pub fn write(&out self, byte: u8, location: usize) {
+        self.0[location] = byte; // currently not possible to index like this, but we could imagine a IndexOut, that will handle this case
+    }
+}
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+**NOTE** This RFC does NOT aim to create new raw pointer types, so no `*out T` or `*uninit T`. There is no point in creating these.
+
+## Rules of `&uninit T`
+
+`&uninit T` should follow some rules in so that is is easy to reason about `&uninit T` locally and maintain soundness
+- `&uninit T` follows the same rules as `&mut T` for the borrow checker
+- `&uninit T` can only be assigned to once
+    - After being written to `&uninit T` are promoted to a `&mut T`
+- Writing does not drop old value.
+    - Otherwise, it would not handle writing to uninitialized memory 
+    - More importantly, dropping requires at least one read, which is not possible with a write-only reference
+- You cannot reference partially initialized memory
+```Rust
+let x: Bar;
+
+fn init_bar(bar: &uninit Bar) { ... }
+fn init_u32(x: &uninit u32) { ... }
+
+x.e = 10.0;
+
+// init_bar(&uninit x); // compile time error: attempting to reference partially initialized memory
+init_u32(&uninit x.d); // fine, x.d is completely uninitialized.
+```
+- Functions and closures that take a `&uninit T` argument must initialize it before returning
+    - You cannot return an `&uninit T`
+    - because of this rule, `T` cannot resolve to `&uninit _` in any case, because it is impossible to parametrically initialize the `&uninit`
+- You can take a `&uninit T` on any `T` that represents uninitialized memory, for example: only the first is ok.
+```Rust
+let x: Foo;
+let y = &uninit x;
```

Comment:
> `let x: Foo; let y: Foo; let z = if cond { &uninit x } else { &uninit y }`

No, because each branch returns an `&uninit` to a different value, and if one branch returns an `&uninit` to a value, all branches must return an `&uninit` to that value.

Reply:
```Rust
let x: Foo;
let y = &uninit x;
// new lines
f(y);
println!("{}", x.b);
```

Because `f(y)` is guarenteed to initialize the value pointed to by `y`, in this case `x`, the print will print the value that the function `f` initialized `x.b` to.